src/Ens.ts

```diff
@@ -0,0 +1,594 @@
+  }
+
+  namehash(domain: string): string {
+    if (!this.checkSupportedDomain(domain)) {
+      throw new ResolutionError(ResolutionErrorCode.UnsupportedDomain, {
+        domain,
+      });
+    }
+    return eip137Namehash(domain);
+  }
+
+  childhash(parentHash: string, label: string): string {
+    return eip137Childhash(parentHash, label);
+  }
+
+  async isSupportedDomain(domain: string): Promise<boolean> {
+    return this.checkSupportedDomain(domain);
+  }
+
+  async owner(domain: string): Promise<string> {
+    const namehash = this.namehash(domain);
+    return await this.callMethod(this.registryContract, 'owner', [namehash]);
+  }
+
+  async resolver(domain: string): Promise<string> {
+    const nodeHash = this.namehash(domain);
+    const resolverAddr = await this.callMethod(
+      this.registryContract,
+      'resolver',
+      [nodeHash],
+    );
+    if (isNullAddress(resolverAddr)) {
+      throw new ResolutionError(ResolutionErrorCode.UnspecifiedResolver);
+    }
+    return resolverAddr;
+  }
+
+  async record(domain: string, key: string): Promise<string> {
+    const returnee = (await this.records(domain, [key]))[key];
+    if (!returnee) {
+      throw new ResolutionError(ResolutionErrorCode.RecordNotFound, {
+        domain,
+        recordName: key,
+      });
+    }
+    return returnee;
+  }
+
+  async records(
+    domain: string,
+    keys: string[],
+  ): Promise<Record<string, string>> {
+    const values = await Promise.all(
+      keys.map(async (key) => {
+        if (key.startsWith('crypto.')) {
+          const ticker = key.split('.')[1];
+          return await this.addr(domain, ticker);
+        }
+        if (key === 'ipfs.html.value' || key === 'dweb.ipfs.hash') {
+          return await this.getContentHash(domain);
+        }
+        const ensRecordName = this.fromUDRecordNameToENS(key);
+        return await this.getTextRecord(domain, ensRecordName);
+      }),
+    );
+    return constructRecords(keys, values);
+  }
+
+  async reverse(
+    address: string,
+    currencyTicker: string,
+  ): Promise<string | null> {
+    if (currencyTicker != 'ETH') {
+      throw new Error(`Ens doesn't support any currency other than ETH`);
+    }
+
+    if (address.startsWith('0x')) {
+      address = address.substr(2);
+    }
+
+    const reverseAddress = address + '.addr.reverse';
+    const nodeHash = this.namehash(reverseAddress);
+    const resolverAddress = await this.resolver(reverseAddress).catch(
+      (err: ResolutionError) => {
+        if (err.code === ResolutionErrorCode.UnspecifiedResolver) {
+          return null;
+        }
+        throw err;
+      },
+    );
+
+    if (isNullAddress(resolverAddress)) {
+      return null;
+    }
+
+    const resolverContract = new EthereumContract(
+      resolverInterface(resolverAddress, EthCoinIndex),
+      resolverAddress,
+      this.provider,
+      this.proxyServiceApiKey,
+    );
+
+    return await this.resolverCallToName(resolverContract, nodeHash);
```

Comment:
we need a final verification, if the `addr` is the same as address from the params then it's a valid reverse resolution. see reference
https://github.com/unstoppabledomains/resolution-go/blob/master/ens.go#L56-L82